trin-core/src/types/accumulator.rs

```diff
@@ -0,0 +1,219 @@
+    /// If current epoch fills up, it will refresh the epoch and add the
+    /// preceding epoch's merkle root to historical epochs.
+    // as defined in:
+    // https://github.com/ethereum/portal-network-specs/blob/e807eb09d2859016e25b976f082735d3aceceb8e/history-network.md#the-header-accumulator
+    fn update_accumulator(&mut self, new_block_header: &Header) {
+        // get the previous total difficulty
+        let last_total_difficulty = match self.current_epoch.header_records.len() {
+            // genesis
+            0 => U256::zero(),
+            _ => {
+                self.current_epoch
+                    .header_records
+                    .last()
+                    .expect("Invalid master accumulator state: No header records for current epoch on non-genesis header.")
+                    .total_difficulty
+            }
+        };
+
+        // check if the epoch accumulator is full
+        if self.current_epoch.header_records.len() == EPOCH_SIZE {
+            // compute the final hash for this epoch
+            let epoch_hash = self.current_epoch.tree_hash_root();
+            // append the hash for this epoch to the list of historical epochs
+            self.historical_epochs
+                .epochs
+                .push(epoch_hash)
+                .expect("Invalid accumulator state, more historical epochs than allowed.");
+            // initialize a new empty epoch
+            self.current_epoch = EpochAccumulator {
+                header_records: HeaderRecordList::empty(),
+            };
+        }
+
+        // construct the concise record for the new header and add it to the current epoch
+        let header_record = HeaderRecord {
+            block_hash: new_block_header.hash(),
+            total_difficulty: last_total_difficulty + new_block_header.difficulty,
+        };
+        self.current_epoch
+            .header_records
+            .push(header_record)
+            .expect("Invalid accumulator state, more current epochs than allowed.");
+    }
+}
+
+/// Datatype responsible for storing all of the merkle roots
+/// for epoch accumulators preceding the current epoch.
+#[derive(Clone, Debug, Decode, Encode)]
+pub struct HistoricalEpochs {
+    epochs: HistoricalEpochsList,
+}
+
+impl TreeHash for HistoricalEpochs {
+    fn tree_hash_type() -> tree_hash::TreeHashType {
+        tree_hash::TreeHashType::List
+    }
+
+    fn tree_hash_packed_encoding(&self) -> Vec<u8> {
+        // in ssz merkleization spec, only basic types are packed
+        unreachable!("List should never be packed.")
+    }
+
+    fn tree_hash_packing_factor() -> usize {
+        // in ssz merkleization spec, only basic types are packed
+        unreachable!("List should never be packed.")
+    }
+
+    fn tree_hash_root(&self) -> tree_hash::Hash256 {
+        // list of composite objects
+        // mix_in_length(merkleize([hash_tree_root(element) for element in value], limit=chunk_count(type)), len(value))
```

Comment:
Same comment as below. This comment is more confusing than helpful to me.

Reply:
Can we remove?

Reply:
I agree that it's useful (and desired) to describe the logic, but I don't think that this comment accomplishes that. If we keep the comment, then can we translate the pseudocode into more of a description?

Reply:
It's not pseudocode, it's legit python. I agree that it would be nice to add more context for folks who haven't been hanging out in py-ssz, though. Something like:
```rs
// When generating the hash root of a list of composite objects, the complete procedure is here
// https://github.com/ethereum/py-ssz/blob/36f3406f814a5e5f4efb059a6928afc2d9d253b4/ssz/sedes/list.py#L113-L129
// Since we know each element is a composite object, we can simplify the python by removing some branches, to:
// mix_in_length(merkleize([hash_tree_root(element) for element in value], limit=chunk_count(type)), len(value))
```